### ISSUE TITLE
IndexOutOfBoundsException in RealConnection

### ISSUE THREAD

**Issue by andylizi** [square/okhttp#4427]
#### Stacktrace
```
java.lang.IndexOutOfBoundsException: Index: 0
	at java.base/java.util.Collections$EmptyList.get(Collections.java:4481)
	at okhttp3.internal.connection.RealConnection.connectTls(RealConnection.java:326)
	at okhttp3.internal.connection.RealConnection.establishProtocol(RealConnection.java:283)
	at okhttp3.internal.connection.RealConnection.connect(RealConnection.java:168)
	at okhttp3.internal.connection.StreamAllocation.findConnection(StreamAllocation.java:257)
	at okhttp3.internal.connection.StreamAllocation.findHealthyConnection(StreamAllocation.java:135)
	at okhttp3.internal.connection.StreamAllocation.newStream(StreamAllocation.java:114)
	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:42)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:126)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:254)
	at okhttp3.RealCall.execute(RealCall.java:92)
```

#### Environment
* OkHttp 3.12.0
* Java 11.0.1 LTS
* Windows 10 64-bit

This bug seems cannot be stably reproduce, so I'm not sure how to write a failing test for this. After I downgraded okhttp to 3.11.0 it works fine.

**Comment by yschimke** [comment]
@andylizi it's also only presumably happening when it fails to verify a host.  Maybe that is what is new

**Comment by andylizi** [comment]
> Are you using some particular setup e.g. Pre-Shared Keys?

No.... I don't think so.
```java
OkHttpClient client = new OkHttpClient.Builder()
        .connectionPool(connectionPool)
        .protocols(Collections.singletonList(Protocol.HTTP_1_1))
        .retryOnConnectionFailure(false)
        .proxySelector(proxySelector)
        .build();

Request request = new Request.Builder()
        .url(url)
        .get()
        .header(HttpHeaders.ACCEPT, "text/html")
        .build();

try (Response response = client.newCall(request).execute()) {
    // ...
}
```

---

#### SSL Server Test Report provided by [ssllabs.com](https://www.ssllabs.com/ssltest/index.html)
![cert-chain](https://user-images.githubusercontent.com/12008103/49366179-a6411800-f723-11e8-9f3d-d9616a90132f.png)
![cipher-suites](https://user-images.githubusercontent.com/12008103/49366182-a6d9ae80-f723-11e8-8832-5a079edf9d3e.png)
![protocol-details](https://user-images.githubusercontent.com/12008103/49366464-5f075700-f724-11e8-90f9-614243aa036b.png)

**Comment by ToxicMushroom** [comment]
I'm also having this issue pls fix

**Comment by MinnDevelopment** [comment]
This seems to be rather potent with version 3.12.0, however that version also contains a fix for an issue of 3.11.0 with proxySelector being null on some JVMs. I'm twisted whether I should be using 3.12.0 with this issue or 3.11.0 with the other. I'd much appreciate a fix for this to void that problem.
